imap_processing/tests/lo/test_lo_l1b.py

```diff
@@ -976,3 +989,297 @@ def test_calculate_histogram_rates_zero_exposure_time(l1b_histrates):
+            "shcoarse": ("epoch", met_times),
+            "num_completed": ("epoch", [28, 28]),
+            "acq_start_sec": ("epoch", met_times),
+            "acq_start_subsec": ("epoch", [0, 0]),
+            "acq_end_sec": ("epoch", met_times),
+            "acq_end_subsec": ("epoch", [0, 0]),
+        },
+        coords={
+            "epoch": epoch_date,
+        },
+    )
+
+    # Mock get_spin_number to return predictable values
+    with patch(
+        "imap_processing.lo.l1b.lo_l1b.get_spin_number", return_value=np.array([0, 28])
+    ):
+        # Act
+        l1b_hist = set_spin_cycle_from_spin_data(l1a_hist, l1b_hist, spin_data)
+
+    # Expected: spin_cycle = spin_start + 7 + (esa_step - 1) * 2
+    # For epoch 0: 0 + 7 + (1-1)*2 = 7, (2-1)*2 = 9, ..., (7-1)*2 = 19
+    # For epoch 1: 28 + 7 + (1-1)*2 = 35, ..., 28 + 7 + (7-1)*2 = 47
+    expected_spin_cycles = np.array(
+        [[7, 9, 11, 13, 15, 17, 19], [35, 37, 39, 41, 43, 45, 47]]
+    )
+
+    # Assert
+    assert "spin_cycle" in l1b_hist.data_vars
+    np.testing.assert_array_equal(l1b_hist["spin_cycle"].values, expected_spin_cycles)
+
+
+def test_set_spin_cycle_from_spin_data_matching_ascs():
+    """Test that science ASCs correctly match to spin ASCs."""
+    # Arrange - Science ASCs that should match different spin ASCs
+    science_met = [100, 200, 300]
+    spin_met = [50, 150, 250]  # Science times fall after these spin times
+
+    epoch_date = met_to_ttj2000ns(science_met)
+    l1a_hist = xr.Dataset(
+        {
+            "hydrogen": (("epoch", "esa_step"), np.zeros((3, 7))),
+        },
+        coords={
+            "epoch": epoch_date,
+            "esa_step": np.arange(1, 8),
+        },
+        attrs={"Logical_source": "imap_lo_l1a_histogram"},
+    )
+
+    l1b_hist = xr.Dataset(coords={"epoch": epoch_date, "esa_step": np.arange(1, 8)})
+
+    spin_data = xr.Dataset(
+        {
+            "shcoarse": ("epoch", spin_met),
+            "num_completed": ("epoch", [28, 28, 28]),
+            "acq_start_sec": ("epoch", spin_met),
+            "acq_start_subsec": ("epoch", [0, 0, 0]),
+            "acq_end_sec": ("epoch", spin_met),
+            "acq_end_subsec": ("epoch", [0, 0, 0]),
+        },
+        coords={"epoch": met_to_ttj2000ns(spin_met)},
+    )
+
+    with patch(
+        "imap_processing.lo.l1b.lo_l1b.get_spin_number",
+        return_value=np.array([0, 28, 56]),
+    ):
+        # Act
+        l1b_hist = set_spin_cycle_from_spin_data(l1a_hist, l1b_hist, spin_data)
+
+    # Assert - Each epoch should use the correct spin start number
+    # science_met[0]=100 -> searchsorted returns 1, -1 = 0 (spin_met[0]=50)
+    # science_met[1]=200 -> searchsorted returns 2, -1 = 1 (spin_met[1]=150)
+    # science_met[2]=300 -> searchsorted returns 3, -1 = 2 (spin_met[2]=250)
+    assert l1b_hist["spin_cycle"][0, 0] == 7  # 0 + 7 + 0
+    assert l1b_hist["spin_cycle"][1, 0] == 35  # 28 + 7 + 0
+    assert l1b_hist["spin_cycle"][2, 0] == 63  # 56 + 7 + 0
```

Comment:
Can you test a different energy step here too?
```suggestion
    assert l1b_hist["spin_cycle"][2, 2] == 67  # 56 + 7 + 2*2
```